README
CPAN::Changes::Dependencies::Details

This module serves as a utility for producing complex change-logs in
"CPAN::Changes" style, specifically tailored to conveying the nature of
dependency changes between releases.

It typically requires some mechanism external to this code to report the
state of prerequisites at given versions, where it computes their
differences and produces sections detailing the kinds of changes you
have elected to itemize.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for CPAN-Changes-Dependencies-Details

0.001008 2017-03-05T07:57:36Z 966bda8

0.001007 2017-03-05T07:11:06Z 6def340
 [Bugfix]
 - Avoid test failures on 5.26 w/ -Ddefault_inc_excludes_dot
 - Closes https://github.com/kentnl/CPAN-Changes-Dependencies-Details/issues/1

 [Dependencies::Stats]
 - Dependencies changed since 0.001006, see misc/*.deps* for details
 - develop: +3 ↑2 (suggests: ↑2)
 - test: ↓1

0.001006 2015-07-19T20:11:38Z adebbd4
 [Bugfix]
 - Guard tests against further changes.

0.001005 2015-07-16T19:01:08Z d3a7714
 [Bugfix]
 - Was too sensitive to whitespace at EOF and was affected by recent changes in CPAN::Changes
 - Now has a guard block in the test so the test will only run when CPAN::Changes itself serializes as expected.
 - Tests were poorly adapted for reporting differences in utf8 data.

 [Dependencies::Stats]
 - Dependencies changed since 0.001004, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +9 ↑2 -3 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -1 (recommends: +1)
 - test: +2 (recommends: ↑1 ↓1)

0.001004 2014-10-11T08:49:42Z 1c27f83
 [Bugfix]
 - Be slightly less dumb how generated releases get added to the log.
 - This is a forwards compatibility fix also ;)

 [Dependencies::Stats]
 - Dependencies changed since 0.001003, see misc/*.deps* for details
 - develop: +2 ↑1 (suggests: ↑1)
 - test: (recommends: ↑1)

0.001003 2014-08-30T01:14:34Z
 [00 Trivial]
 - no code changes.
 - Fixup author metadata

 [Dependencies::Stats]
 - Dependencies changed since 0.001002, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)

 [Tests]
 - Slight coverage improvement

0.001002 2014-08-15T21:12:19Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - whitespace adjustment.

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - develop: +1 (suggests: ↑1)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

0.001001 2014-08-06T19:21:21Z
 [00 Trivial]
 - Update ReportPrereqs test
 - no code changes

 [Dependencies::Stats]
 - Dependencies changed since 0.001000, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - test: -3

0.001000 2014-08-01T11:36:44Z
 - First version.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Create CPAN::Changes style file only containing dependency change information",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "CPAN-Changes-Dependencies-Details",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "CPAN::Changes::Dependencies::Details",
  "PREREQ_PM" => {
    "CPAN::Changes" => 0,
    "CPAN::Changes::Group::Dependencies::Details" => "0.001001",
    "CPAN::Changes::Release" => 0,
    "Carp" => 0,
    "Moo" => 0,
    "MooX::Lsub" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Differences" => 0,
    "Test::More" => 0,
    "charnames" => 0
  },
  "VERSION" => "0.001008",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::Changes" => 0,
  "CPAN::Changes::Group::Dependencies::Details" => "0.001001",
  "CPAN::Changes::Release" => 0,
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moo" => 0,
  "MooX::Lsub" => 0,
  "Test::Differences" => 0,
  "Test::More" => 0,
  "charnames" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/CPAN/Changes/Dependencies/Details.pm
use 5.006;    # our
use strict;
use warnings;

package CPAN::Changes::Dependencies::Details;

our $VERSION = '0.001008';

# ABSTRACT: Create CPAN::Changes style file only containing dependency change information

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moo qw( extends around );
use MooX::Lsub qw( lsub );
use Carp qw( croak );
use CPAN::Changes::Release;
use CPAN::Changes::Group::Dependencies::Details 0.001001;    # First useful version

extends 'CPAN::Changes';

lsub change_types => sub { [qw( Added Changed Removed )] };
lsub phases       => sub { [qw( configure build runtime test )] };
lsub types        => sub { [qw( requires )] };

sub load        { croak 'This module can only generate dependency details, not read them' }
sub load_string { croak 'This module can only generate dependency details, not read them' }

my $release_keys = [ 'changes',     'version',     'date',         'note', ];
my $group_keys   = [ 'new_prereqs', 'old_prereqs', 'prereqs_diff', 'all_diffs', ];

sub _mk_release {
  my ( $self, $release ) = @_;
  my $input_args   = { %{$release} };
  my $release_args = {};
  my $group_args   = {};

  for my $release_key ( @{$release_keys} ) {
    next unless exists $input_args->{$release_key};
    $release_args->{$release_key} = delete $input_args->{$release_key};
  }

  for my $group_key ( @{$group_keys} ) {
    next unless exists $input_args->{$group_key};
    $group_args->{$group_key} = delete $input_args->{$group_key};
  }

  my $release_object = CPAN::Changes::Release->new( %{$release_args} );

  for my $change_type ( @{ $self->change_types } ) {
    for my $phase ( @{ $self->phases } ) {
      for my $type ( @{ $self->types } ) {
        my $group = CPAN::Changes::Group::Dependencies::Details->new(
          change_type => $change_type,
          phase       => $phase,
          type        => $type,
          %{$group_args},
        );
        next unless $group->has_changes;
        $release_object->attach_group($group);
      }
    }
  }
  return $release_object;
}

around add_release => sub {
  my ( $orig, $self, @releases ) = @_;
  for my $release (@releases) {
    my $release_object = $self->_mk_release($release);
    $self->$orig($release_object);
  }
  return;
};

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

CPAN::Changes::Dependencies::Details - Create CPAN::Changes style file only containing dependency change information

=head1 VERSION

version 0.001008

=head1 SYNOPSIS

  use CPAN::Changes::Dependencies::Details;
  my $details = CPAN::Changes::Dependencies::Details->new(
    preamble     => "Some message",
    change_types => [qw( Added Changed Removed )],
    phases       => [qw( build configure runtime test )],
    types        => [qw( requires recommends )],
  );

  $changes->add_release({
    version     => '0.002',
    date        => '2009-07-06',
    old_prereqs => CPAN::Meta->load_file('Dist-Foo-0.001/META.json')->effective_prereqs,
    new_prereqs => CPAN::Meta->load_file('Dist-Foo-0.002/META.json')->effective_prereqs,
  });

  print $changes->serialize;

=head1 DESCRIPTION

This module serves as a utility for producing complex change-logs in C<CPAN::Changes> style,
specifically tailored to conveying the nature of dependency changes between releases.

It typically requires some mechanism external to this code to report the state of prerequisites
at given versions, where it computes their differences and produces sections detailing
the kinds of changes you have elected to itemize.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
